src/wallet/spend.cpp

```diff
@@ -626,11 +657,13 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, AvailableCoins
             if (!fRejectLongChains) {
                 if (auto r8{AttemptSelection(wallet, value_to_select,
                                       CoinEligibilityFilter(0, 1, std::numeric_limits<uint64_t>::max(), std::numeric_limits<uint64_t>::max(), true /* include_partial_groups */),
-                                      coins, coin_selection_params)}) {
+                                      coins, coin_selection_params, /*allow_mixed_output_types=*/true)}) {
                     return r8;
                 }
             }
         }
+        // If our wallet does not allow us to spend unconfirmed change, attempt to fund the transaction with mixed OutputTypes
+        if (auto r9{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 1, 0), coins, coin_selection_params, /*allow_mixed_output_types=*/true)}) return r9;
```

Comment:
my reasoning for starting it at the `r6` back-off is because that is also the first back-off where we allow partial spending of `OutputGroups`. so the status quo in the wallet is "spend unconfirmed change before relaxing privacy-preserving constraints."

you could argue that mixed outputs are not as bad as partially spending an output group w.r.t privacy, but i would still advocate for it being considered along with unconfirmed change, meaning if spending unconfirmed change is more efficient in terms of the waste metric, do that instead of mixing.

Reply:
I'm not sure that's an apt comparison. Address reuse is generally discouraged and hopefully happens seldom for most wallets. In addition, we already run opportunistic avoid-partial-spend in parallel for every single coin selection attempt and prefer its result as long as it is not more expensive even when not explicitly required.

This PR however follows a change in the wallet behavior that will generally match the output type of the recipient on every spend. We must expect wallets to frequently have outputs of different types as a result. This PR does not introduce avoidance of mixing as parallel coin selection results but rather requires it for almost all selection attempts. Splitting the UTXO pool into two or more sets doesn't half the size of the possible input set candidates, but reduces the combination space to tiny fractions.

I expect this already to cause a reduction in changeless transactions, but worse, I expect the change as proposed to considerably increase the use of unconfirmed inputs and thusly reduce the resulting transactions' reliability. Given the existing feerate estimation issue with unconfirmed inputs, this would more frequently cause the wallet to undershoot the target feerate, especially if we allow up to 12 unconfirmed inputs before considering mixing.

I'd prefer that mixing be allowed with r2, I'd be okay with r3. I might be convinced to introduce mixing on r4 via simulation results showing that coin selection results and reliability do not get significantly worse. My stance on this would generally be improved by us having finished ancestor aware fee estimation and the waste metric having been updated to account for automatic bumping of low feerate parent transactions.

I'd also be curious what others think about this.

Reply:
Let me add an example to illustrate my point.

Five UTXOs produce 31 non-empty combinations:
![image](https://user-images.githubusercontent.com/4060799/167263471-ebac8d8b-7699-411d-9128-422b626fabdb.png)

Two sets of three and two UTXOs only yield 7+3 combinations:
![image](https://user-images.githubusercontent.com/4060799/167263547-4929efa7-c83c-49ed-8ccb-53842f7e9d5c.png)
![image](https://user-images.githubusercontent.com/4060799/167263566-2255c81e-84a5-45a6-9242-b3ccfd708cf7.png)

Reply:
thanks for the detailed explanation and example. very good points about splitting the UTXO set. 

I've changed the logic and added a new back-off that allows mixing _before_ considering unconfirmed change. as such, you will still get the privacy improvement when spending from confirmed UTXOs. 

what's nice about how this is implemented is that we can easily move the logic up or down by changing the boolean so if we are able to fix the fee estimation issue and convince ourselves via simulations that there are not undesirable effects, we can easily move the restriction down at a later date by changing a boolean.